bd-report-parsers/src/javascript.rs

```diff
@@ -0,0 +1,407 @@
+    .map(|m| builder.create_string(m));
+
+  let mut device_info = v_1::DeviceMetricsArgs {
+    platform: device_metrics.platform,
+    manufacturer: manufacturer_offset,
+    model: model_offset,
+    time: Some(&timestamp),
+    ..Default::default()
+  };
+
+  if device_metrics.os_version.is_some()
+    || device_metrics.os_brand.is_some()
+    || device_metrics.os_kernversion.is_some()
+  {
+    let os_version_offset = device_metrics
+      .os_version
+      .as_ref()
+      .map(|v| builder.create_string(v));
+    let os_brand_offset = device_metrics
+      .os_brand
+      .as_ref()
+      .map(|b| builder.create_string(b));
+    let os_kernversion_offset = device_metrics
+      .os_kernversion
+      .as_ref()
+      .map(|k| builder.create_string(k));
+    let os_build = v_1::OSBuild::create(
+      &mut builder,
+      &v_1::OSBuildArgs {
+        version: os_version_offset,
+        brand: os_brand_offset,
+        kern_osversion: os_kernversion_offset,
+        ..Default::default()
+      },
+    );
+    device_info.os_build = Some(os_build);
+  }
+
+  if let Some(arch) = device_metrics.architecture {
+    device_info.arch = arch;
+  }
+
+  let mut app_info = v_1::AppMetricsArgs {
+    app_id: app_metrics
+      .app_id
+      .as_ref()
+      .map(|id| builder.create_string(id)),
+    version: app_metrics
+      .version
+      .as_ref()
+      .map(|v| builder.create_string(v)),
+    javascript_engine: app_metrics.javascript_engine,
+    ..Default::default()
+  };
+
+  if let Some(version_code) = app_metrics.version_code {
+    let build_number = v_1::AppBuildNumber::create(
+      &mut builder,
+      &v_1::AppBuildNumberArgs {
+        version_code,
+        ..Default::default()
+      },
+    );
+    app_info.build_number = Some(build_number);
+  }
+
+  let args = v_1::ReportArgs {
+    type_: report_type,
+    errors: Some(builder.create_vector(&[error])),
+    app_metrics: Some(v_1::AppMetrics::create(&mut builder, &app_info)),
+    device_metrics: Some(v_1::DeviceMetrics::create(&mut builder, &device_info)),
+    ..Default::default()
+  };
+  let report_offset = v_1::Report::create(&mut builder, &args);
+
+  builder.finish(report_offset, None);
+  builder.finished_data().to_vec()
+}
+
+pub fn build_javascript_error_report_to_file(
+  error_name: &str,
+  error_message: &str,
+  stack_trace: &str,
+  is_fatal: bool,
+  debugger_id: Option<&str>,
+  timestamp_seconds: u64,
+  timestamp_nanos: u32,
+  device_metrics: &JavaScriptDeviceMetrics,
+  app_metrics: &JavaScriptAppMetrics,
+  destination_path: &str,
+) -> anyhow::Result<()> {
+  let bytes = build_javascript_error_report(
+    error_name,
+    error_message,
+    stack_trace,
+    is_fatal,
+    debugger_id,
+    timestamp_seconds,
+    timestamp_nanos,
+    device_metrics,
+    app_metrics,
+  );
+  std::fs::write(destination_path, bytes)
+    .map_err(|e| anyhow::anyhow!("failed to write report to file: {e}"))?;
+  Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
```

Comment:
We typically leave the tests in a separate file, you'll see that pattern all over